### PR TITLE
taskgroup decorator (#11870)

### LINKED ISSUE #11870: Add @taskgroup decorator
**Description**

Similarly to `@dag` and `@task` decorators, we can add a decorator to easily generate `TaskGroup` instances. 

**Use case / motivation**

This would be used for reusable pieces of DAGs. An example may be to hyperparameter tune an ML pipeline. You may want to generate several parallel subpipelines with different settings/parameters to be executed in parallel. At the end pull the result and decide the best parametrization.  

```py
@task
collect_dataset(...)
  pass

@task
def train_model(...)
   pass

@task_group
def movielens_model_pipeline(learning_rate: int, feature_list: list, dataset: XComArg):  
  dataset = filter_dataset(dataset, feature_list)
  train_model(dataset, learning_rate)

@dag
def movielens_hpt(dataset_path: str, feature_list:list=['year', 'director']):
  dataset = load_dataset(dataset_path)
  for i in range(0.1, 0.9, step=0.1):
    movielens_model_pipeline(i, feature_list, dataset)
  
  decide_best_model(...)
```

**Related Issues**

https://github.com/apache/airflow/issues?q=is%3Aopen+is%3Aissue+label%3AAIP-31

### PR DESCRIPTION
**Pull request by VBhojawala**
Added taskgroup decorator which can be used to create taskgroup from python callable. Inside python callable tasks can be grouped by calling tasks (python callable ) created with task decorator.

- taskgroup decorator takes both optional and keyword argument which are passed to Constructor of TaskGroup class.

- TaskGroup can be created with one of following syntax

``` python
@taskgroup
@taskgroup()
@taskgroup(group_id='group_name')
```

- TaskGroup class  constructor takes one mandatory argument group_id, if not given in decorator it sets group_id to python callable name.

Following is a simple example demonstrating use of taskgroup decorator grouping multiple tasks.

``` python
@task
def task_1(value):
  return f'[ Task1 {value} ]'


@task
def task_2(value):
  print(f'[ Task2 {value} ]')


@taskgroup
def section_1(value):
  return task_2(task_1(value))
```

taskgroup decorator utilizes existing TaskGroup context manager currently used for creating TaskGroup, which means we can create nested taskgroup by created nested callable. Following is an example demonstrating use of nested taskgroup.

``` python
@task
def task_start():
    return '[Task_start]'

@task
def task_end():
    print(f'[ Task_End ]')

@task
def task_1(value):
    return f'[ Task1 {value} ]'

@task
def task_2(value):
    print(f'[ Task2 {value} ]')

@task
def task_3(value):
    return f'[ Task3 {value} ]'

@task
def task_4(value):
    print(f'[ Task4 {value} ]')

@taskgroup
def section_1(value):

    @taskgroup
    def section_2(value2):
        return task_4(task_3(value2))

    op1 = task_2(task_1(value))
    return section_2(op1) 
```

Dedicated test cases for taskgroup decorator is created in file 
/tests/utils/test_task_group_decorator.py

closes: #11870

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/utils/task_group.py [hunk 1]
```diff
@@ -344,3 +346,51 @@ def get_current_task_group(cls, dag: Optional["DAG"]) -> Optional[TaskGroup]:
...
+            with TaskGroup(*task_group_bound_args.args, **task_group_bound_args.kwargs) as tg_obj:
+                # Invoke function to run Tasks inside the TaskGroup
+                f(**f_sig.arguments)
+
+            # Return task_group object such that it's accessible in Globals.
+            return tg_obj
+
+        return cast(T, factory)
+
+    if callable(python_callable):
```
Comment (yuqian90) [inline airflow/utils/task_group.py]: I think this is the same as providing just a `tg_kwargs`? If there's no `*tg_args`, the user can't put in positional arguments after the callable. You shouldn't need to explicitly raise an `AirflowException` for this.

```python
def taskgroup(python_callable: Optional[Callable], **tg_kwargs)
```

File: tests/utils/test_task_group_decorator.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        'section_1.task_2',
+        'section_1.section_2',
+    }
+
+    # Testing TaskGroup consisting Tasks created using task decorator
+    assert dag.task_dict['task_start'].downstream_task_ids == {'section_1.task_1'}
+    assert dag.task_dict['section_1.task_2'].downstream_task_ids == {'section_1.section_2.task_3'}
+    assert dag.task_dict['section_1.section_2.task_4'].downstream_task_ids == {'task_end'}
+
+    def extract_node_id(node, include_label=False):
```
Comment (yuqian90) [inline tests/utils/test_task_group_decorator.py]: I suggest putting these tests inside the existing `test_task_group.py` instead so that things like `extract_node_id` can be reused. It also makes sense because they are closely related.

File: tests/utils/test_task_group_decorator.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    # Testing Tasks ing DAG
+    assert set(dag.task_group.children.keys()) == {'section_1', 'task_start', 'task_end'}
+    assert set(dag.task_group.children['section_1'].children.keys()) == {
+        'section_1.task_2',
+        'section_1.task_3',
+        'section_1.task_1',
+    }
+
+    # Testing Tasks downstream
+    assert dag.task_dict['task_start'].downstream_task_ids == {'section_1.task_1'}
```
Comment (yuqian90) [inline tests/utils/test_task_group_decorator.py]: Maybe also mix `TaskGroup` context manager and decorator in the same DAG to test if they play nice together?

**Review by yuqian90 — COMMENTED** [review]
Thanks for the updates.

By the way, you don't have to open a new PR to make these changes. When you do this the original history of the comments are left on the old PR.

**Comment by VBhojawala** [comment]
> Thanks for the updates.
> 
> By the way, you don't have to open a new PR to make these changes. When you do this the original history of the comments are left on the old PR.

Earlier I raised pull request from Fork. In order to work on multiple issue simultaneously i closed old PR and raised from branch of fork.

File: airflow/utils/task_group.py (hunk 1 above)
Comment (VBhojawala) [reply airflow/utils/task_group.py]: Yes, i will remove elif part.

File: tests/utils/test_task_group_decorator.py (hunk 2 above)
Comment (VBhojawala) [reply tests/utils/test_task_group_decorator.py]: Ok, i will move it to test_task_group.py

File: tests/utils/test_task_group_decorator.py (hunk 3 above)
Comment (VBhojawala) [reply tests/utils/test_task_group_decorator.py]: I made suggested changes and added a new Dag for testing  TaskGroup context manager (section1) with nested TaskGroup decorator (section2).
Screenshot from 2020-11-10 15-29-47

**Review by VBhojawala — COMMENTED** [review]
Added Suggested changes and new test case Dag for nested taskgroup decorator and TaskGroup context manager.

**Comment by VBhojawala** [comment]
I think label attached by bot  ``area:Docs``  is not accurate for this PR. can it be corrected?

File: tests/utils/test_task_group.py [hunk 4]
```diff
@@ -559,3 +560,239 @@ def test_duplicate_group_id():
...
+        """ Dummy Task2"""
+        return f'[ Task2 {value} ]'
+
+    @task
+    def task_3(value):
+        """ Dummy Task3"""
+        print(f'[ Task3 {value} ]')
+
+    # Creating TaskGroups
+    @taskgroup(group_id='section_1')
```
Comment (casassg) [inline tests/utils/test_task_group.py]: Maybe change name of group_id or function to make sure overwrite is working properly

**Review by casassg — CHANGES_REQUESTED** [review]
Let's make sure this is importable from airflow.decorators similarly to `@dag` and `@task`

File: tests/utils/test_task_group.py (hunk 4 above)
Comment (VBhojawala) [reply tests/utils/test_task_group.py]: Thank you for the suggestions, i have added taskgroup to ``decorators.py`` and updated test case as suggested.

**Comment by ashb** [comment]
Does this actually need to be a decorator? How about this:

```python
@task
collect_dataset(...)
  ...

@task
def train_model(...)
   ...

@task_group
def movielens_model_pipeline(learning_rate: int, feature_list: list, dataset: XComArg):  
  dataset = filter_dataset(dataset, feature_list)
  train_model(dataset, learning_rate)

@dag
def movielens_hpt(dataset_path: str, feature_list:list=['year', 'director']):
  dataset = load_dataset(dataset_path)

  # NEW HERE
  with dag.taskgroup():
     for i in range(0.1, 0.9, step=0.1):
       movielens_model_pipeline(i, feature_list, dataset)
  
  decide_best_model(...)
```

Or nested groups:

```python
with dag.taskgroup() as group
  with group.taskgroup(name="Inner group"):
     for i in range(0.1, 0.9, step=0.1):
       movielens_model_pipeline(i, feature_list, dataset)
```

What do you think @VBhojawala @casassg ?

**Comment by casassg** [comment]
@ashb Ideally you get a new task group for every invocation of `movielens_model_pipeline` in order to get it grouped in the UI. Not sure how this relates to the examples. I see nested taskgroups in the example you mention.

**Comment by ashb** [comment]
I think my example would do that too.

I haven't explored if it's possible, but it should be - to me task groups didn't feel like a decorator, but more of a context manager.

We could do both if we want

**Comment by casassg** [comment]
You mean something like:

```
@dag
def movielens_hpt(dataset_path: str, feature_list:list=['year', 'director']):
  dataset = load_dataset(dataset_path)

  # NEW HERE
  for i in range(0.1, 0.9, step=0.1):
    with dag.taskgroup(f'pipeline_{i}'):
       movielens_model_pipeline(i, feature_list, dataset)
  
  decide_best_model(...)
```

I think this could work. I think it would be fun to have it be generated automatically for you, but agree that both approaches make sense.

**Comment by ashb** [comment]
Yeah.

Thinking more about the decorator form:

```
@taskgroup
def section(value):
```

For that to need to be a function, it probably needs to be able to override the TG name i.e. this usage:

```python
    for i in ...:
        section(name=f'section_{i}')
```

(We could of course have the same behaviour as for tasks, where we auto-apply a `__1` suffix etc. This could be what we do, I haven't look at the code PR yet, sorry!)

File: airflow/utils/task_group.py [hunk 5]
```diff
@@ -344,3 +357,52 @@ def get_current_task_group(cls, dag: Optional["DAG"]) -> Optional[TaskGroup]:
...
+    :param tg_args: Arguments for TaskGroup object
+    :type tg_args: list
+    :param tg_kwargs: Kwargs for TaskGroup object.
+    :type tg_kwargs: dict
+    """
+
+    def wrapper(f: T):
+        # Setting group_id as function name if not given
+        if len(tg_args) == 0 and 'group_id' not in tg_kwargs.keys():
+            tg_kwargs['group_id'] = f.__name__
```
Comment (casassg) [inline airflow/utils/task_group.py]: Can we add logic to autogenerate group_id if it already exists on the DAG? 

A taskgroup decorated function can be invoked multiple times, this should generate multiple taskgroups. 

We can follow a similar approach than @task decorator where we append `__1` to the end. Probs we can reuse the same regex.

File: airflow/utils/task_group.py [hunk 6]
```diff
@@ -344,3 +357,52 @@ def get_current_task_group(cls, dag: Optional["DAG"]) -> Optional[TaskGroup]:
...
+
+        # Get dag initializer signature and bind it to validate that task_group_args,
+        # and task_group_kwargs are correct
+        task_group_sig = signature(TaskGroup.__init__)
+        task_group_bound_args = task_group_sig.bind_partial(*tg_args, **tg_kwargs)
+
+        @functools.wraps(f)
+        def factory(*args, **kwargs):
```
Comment (casassg) [inline airflow/utils/task_group.py]: May be nice to provide a specific kwarg for overwriting the group_id

**Review by casassg — COMMENTED** [review]
Agree. I would suggest the following:

- Accept a reserved keyword (ex: `group_id=f'section_{i}'`) when calling the decorated function to overwrite the TaskGroup group_id. 
- Automatically geneate a suffix similar to task decorator.

It's currently not implemented.  Thoughts Ash?

File: airflow/utils/task_group.py (hunk 5 above)
Comment (VBhojawala) [reply airflow/utils/task_group.py]: Sure, can you guide me how to get list of existing group_id inside taskgroup decorator?

File: airflow/utils/task_group.py (hunk 6 above)
Comment (VBhojawala) [reply airflow/utils/task_group.py]: Currently for TaskGroup constructor group_id is by default first positional argument or it can be keyword argument.
Same can be done with current implementation of taskgroup decorator.

Are you suggesting to place group_id as optional keyword argument in decorator signature?

File: airflow/utils/task_group.py (hunk 5 above)
Comment (VBhojawala) [reply airflow/utils/task_group.py]: I have added auto generate group_id if it already exists  same as task. and also added test case for it.

**Comment by casassg** [comment]
I believe you may have rebased on the wrong commit, seems the PR now includes changes from upstream.

**Comment by VBhojawala** [comment]
oops, re-raising pull request.
